wp-content/plugins/tapestry/controller/class.tapestry-controller.php

```diff
@@ -3,4 +3,83 @@
  * Add/update/retrieve Tapestry post and its child nodes
  * 
  */
-class TapestryController {}
\ No newline at end of file
+class TapestryController {
+    /**
+     * Retrieve a Tapestry post
+     * 
+     * @param  Number @postId The Tapestry postId
+     * @return Object Tapestry
+     */
+    public function getTapestry($postId = null) {
+        // TODO: Use $this->postId that's passed in the constructor
+        // TODO: uncomment the two lines below for error handling
+        // after the PR for saving tapestry is merged.
+        /*
+        if (!$this->_isValidPostId($postId)) {
+            return $this->throwsError('INVALID_POST_ID');
+        }
+        */
+        $tapestry = $this->_getTapestryById($postId);
+        return $tapestry;
+    }
+
+    private function _getTapestryById($postId) {
+        $tapestry = get_post_meta($postId, 'tapestry', true);
+
+        $metadatas = array_map(function($nodeId) {
+            return get_metadata_by_mid('post', $nodeId);
+        }, $tapestry->nodes);
+
+        $nodeDatas = array_map(function($metadata) {
+            $postId = $metadata->meta_value->post_id;
+            $nodeData = get_post_meta($postId, 'tapestry_node_data', true);
+            return $this->_updateNodeData($nodeData, $metadata);
+        }, $metadatas);
+
+        $tapestry->nodes = $nodeDatas;
+
+        // TODO: delete the below when being able to create tapestry from scratch
+        $tapestry->links = $this->_getNewLinks($tapestry->links, $nodeDatas);
+        return $tapestry;
+    }
+
+    private function _updateNodeData($nodeData, $metadata) {
+        // Update node data here to match its own version
+        // This enables the same node to have multiple versions
+        $nodeData->id = (int) $metadata->meta_id;
+        $nodeData->title = $metadata->meta_value->title;
+        $nodeData->fx = $metadata->meta_value->coordinates->x;
+        $nodeData->fy = $metadata->meta_value->coordinates->y;
+        return $nodeData;
+    }
+
+    // TODO: Remove this when we can build a tapestry from scratch
+    // HACK - create a new links array that works with new IDs
+    private function _getNewLinks($oldLinks, $nodes) {
```

Comment:
Instead of doing this, could you manually update the data in the database to contain the actual node IDs? This is too hacky to go into master.

Reply:
The function is meant to support Kevin's work (I don't need this function at all). Yes, if he agrees to do that manually...

Reply:
Also, I think the data in the DB is serialized when inserted. It might be tricky to modify part of the data.

Reply:
Ok if it's too time-consuming, you don't have to do it. But if you wanted to do it, you can just give Kevin and I an export of your DB `posts` and `postmeta` tables. And changing it should not be difficult - you can just unserialize the data by googling "unseralize php" and using any of those tools and then seralizing it again before inserting it into the db. Again, if you think it's too time-consuming, then just leave it and we can remove it from the code later.